src/hotspot/share/runtime/os.cpp

```diff
@@ -1745,13 +1745,30 @@ void os::print_memory_mappings(outputStream* st) {
   os::print_memory_mappings(nullptr, (size_t)-1, st);
 }
 
+// Pretouching must use a store, not just a load.  On many OSes loads from
+// fresh memory would be satisfied from a single mapped page containing all
+// zeros.  We need to store something to each page to get them backed by
+// their own memory, which is the effect we want here.  An atomic add of
+// zero is used instead of a simple store, allowing the memory to be used
+// while pretouch is in progress, rather than requiring users of the memory
+// to wait until the entire range has been touched.  This is technically
+// a UB data race, but doesn't cause any problems for us.
 void os::pretouch_memory(void* start, void* end, size_t page_size) {
-  for (volatile char *p = (char*)start; p < (char*)end; p += page_size) {
-    // Note: this must be a store, not a load. On many OSes loads from fresh
-    // memory would be satisfied from a single mapped page containing all zeros.
-    // We need to store something to each page to get them backed by their own
-    // memory, which is the effect we want here.
-    *p = 0;
+  assert(is_power_of_2(page_size), "page size misaligned: %zu", page_size);
+  assert(page_size >= sizeof(int), "page size too small: %zu", page_size);
+  if (start < end) {
+    // We're doing concurrent-safe touch and memory state has page
+    // granularity, so we can touch anywhere in a page.  Touch at the
+    // beginning of each page to simplify iteration.
+    char* cur = static_cast<char*>(align_down(start, page_size));
+    void* last = align_down(static_cast<char*>(end) - 1, page_size);
+    assert(cur <= last, "invariant");
+    // Iterate from first page through last (inclusive), being careful to
+    // avoid overflow if the last page abuts the end of the address range.
+    for ( ; true; cur += page_size) {
+      Atomic::add(reinterpret_cast<int*>(cur), 0, memory_order_relaxed);
+      if (cur == last) break;
+    }
```

Comment:
Took me some time to check the loop. Suggestion:

```
    do {
      Atomic::add(reinterpret_cast<int*>(cur), 0, memory_order_relaxed);
      cur += page_size;
    } while (cur < last);
```

Reply:
Intentionally didn't write it that way, because of the potential for UB pointer arithmetic overflow in `cur += page_size`.

Reply:
Also, that won't touch the last page when first page != last page, unless the last page is made exclusive, which calculation could also suffer from UB pointer arithmetic overflow.  Of course, the caller could suffer from that too; we can't so easily fix that.  Representing address ranges as [start, end) just generally suffers from that problem.

Reply:
All right, what irked me originally is that `cur == last` break assumes both `cur` and `last` are multiples of `page_size`, otherwise this loops "indefinitely" (until successive overflows help, I guess). This is true now, but it is not even asserted. Maybe it should be `cur >= last`?

Reply:
I assumed that having both values be initialized as page-aligned just
before the loop made that obvious.

A slightly different way to write it might be:
```
void os::pretouch_memory(void* start, void* end, size_t page_size) {
  assert(start <= end, "invalid range: " PTR_FORMAT " -> " PTR_FORMAT, p2i(start), p2i(end));
  assert(is_power_of_2(page_size), "page size misaligned: %zu", page_size);
  assert(page_size >= sizeof(int), "page size too small: %zu", page_size);
  if (start < end) {
    // We're doing concurrent-safe touch and memory state has page
    // granularity, so we can touch anywhere in a page.  We need int
    // alignment for the touch.
    int* cur = static_cast<int*>(align_down(start, sizeof(int)));
    void* last = align_down(static_cast<char*>(end) - 1, page_size);
    assert(cur <= last, "invariant");
    // Iterate from first page through last (inclusive), being careful to
    // avoid overflow if the last page abuts the end of the address range.
    for ( ; true; cur += (page_size / sizeof(int))) {
      Atomic::add(cur, 0, memory_order_relaxed);
      if (cur >= last) break;
    }
  }
}
```
I don't know if you'd find that clearer.

Reply:
`int* cur` does not make it cleaner. I am fine with `align_down`-ing both `cur` and `last` to `page_size`, and then using `cur >= last` for extra safety. In other words, just change `==` to `>=` in the breaking condition.

Reply:
Done.